### PR TITLE
Clean up CI

### PR DESCRIPTION
**Pull request by blink1073**
- Improve test matrix
- Add concurrency limit handling for GitHub Actions
- Add cron job to run tests daily
- Use base-setup from maintainer-tools to handle python version, caching, and pip upgrades
- Add tests for minimum versions, pre-releases, and sdist testing
- Update pytest config and handle warnings

### CODE REVIEW

**Comment by blink1073** [comment]
cc @kevin-bates, mind taking a look at this one?  I'd like to make a release of `jupyter_core` tomorrow.

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -15,42 +23,76 @@ jobs:
...
-          # pywin32 not available
+        python-version: ["3.7", "3.10"]
+        include:
           - os: windows-latest
-            python-version: "3.10"
+            python-version: "3.9"
+          - os: ubuntu-latest
+            python-version: "pypy-3.8"
+          - os: macos-latest
+            python-version: "3.8"
```
Comment (kevin-bates) [inline .github/workflows/test.yml:33]: :+1:

**Review by kevin-bates — APPROVED** [review]
LGTM - thanks @blink1073!
